xds/internal/xdsclient/client.go

```diff
@@ -250,8 +250,10 @@ type InboundListenerConfig struct {
 // RouteConfigUpdate contains information received in an RDS response, which is
 // of interest to the registered RDS watcher.
 type RouteConfigUpdate struct {
+	// Name is the RDS resource name for this response, which is specified in
```

Comment:
Used here by rds handler. Since on the server side, there can be multiple filter chains, each with a dynamic RDS specified. Thus, I coded a rds handler to handle everything in another PR I assigned you: https://github.com/grpc/grpc-go/pull/4629/files#diff-7d577e44fa077317c431482dc9284d996bbc87f53bda2d69d73c9cf1e914435eR104. Me and Menghan discussed this last week. The two options were making this a closure in rds_handler, or adding RouteName. I tried both, and RouteName was a lottttt cleaner, and also I think it might be used in other situations going forward.

Reply:
Why is the closure option not clean? I find adding the `RouteName` here to be not clean because it is not needed on the client-side.

Reply:
First, the closure needed to close on a loop variable so the memory that it would refer to was always the last loop variable, and it added a lot of weirdness to. Second, it just made this part of my code: https://github.com/grpc/grpc-go/pull/4629/files#diff-7d577e44fa077317c431482dc9284d996bbc87f53bda2d69d73c9cf1e914435eR70 look really ugly, and I also think it makes sense to add the name to the update. It feels like what happens with ClusterName (even though, you're right this is only on client side, not both), and it just feels like a consequential thing to persist.

Reply:
Also, I talked to Menghan about this, and he mentioned he was indifferent and I could choose.

Reply:
Doesn't something like this work?

```golang
for routeName := range routeNamesToWatch {
  if _, ok := rh.rdsCancels[routeName]; !ok {
    func(routeName string) {
      rh.rdsCancels[routeName] = rh.parent.xdsC.WatchRouteConfig(routeName, func(update xdsclient.RouteConfigUpdate, err error) {
        rh.handleRouteUpdate(routeName, update, err)
      }
    }(routeName)
  }
}

func (rh *rdsHandler) handleRouteUpdate(routeName string, update xdsclient.RouteConfigUpdate, err error) {
}
```

Reply:
We *could* have done without putting the `ClusterName` in the `ClusterUpdate` as well. And now it is part of so many tests and it is hard to fix. Without a good enough reason to have the `RouteName` in the `RouteConfigUpdate`, I prefer that we don't put it in there.

Reply:
IIRC, there was some usage of the clustername, but I forgot what it was. Interesting, so RouteName is essentially going to be a maintenance problem moving forward as it touches so many fields. So I trust your code works, and if it does, I think I see what I was missing. I had essentially exactly what you had now, expect rather than have an anonymous function, I tried both a. the routeName variable, and b. a local variable. What I suspect happened is the memory did not evaluate to the desired memory, as that memory would represent the last iteration. The anonymous function calls it with routeName, which means the routeName variable being closed on will be the desired one. Will work on merging this with your rds handler change tomorrow.

Reply:
That doesn't work, same problem as I had when I did it before. It doesn't have the correct RouteName.

Reply:
Got rds handler to work with closure. Deleting this now.